File: common/deplacement_supervisor/curve.py
class Curve:
	def __init__(self, departure_speed: float, max_speed: float, arrival_speed: float, total_distance: float, acceleration_max: float, decceleration_max: float) -> None:
		# We split the movement in 3 phases: departure, max speed, arrival

		self.departure_speed = departure_speed
		self.mid_speed = max_speed
		self.arrival_speed = arrival_speed

		self.total_distance = total_distance
		self.departure_distance = (max_speed ** 2 - departure_speed ** 2) / (2 * acceleration_max)  # Departure total_distance
		self.arrival_distance = (max_speed ** 2 - arrival_speed ** 2) / (2 * decceleration_max)  # Arrival total_distance
		self.mid_distance = self.total_distance - self.departure_distance - self.arrival_distance

		self.departure_time = 2 * self.departure_distance / (self.departure_speed + self.mid_speed)  # total time of departure phase
		self.arrival_time = 2 * self.arrival_distance / (self.arrival_speed + self.mid_speed)  # total time of arrival phase
		self.mid_time = self.mid_distance / self.mid_speed

		self.total_time = self.departure_time + self.mid_time + self.arrival_time

		self.Cd = (self.mid_speed - self.departure_speed) / self.departure_time  # Departure phase slope
		self.Ca = (self.arrival_speed - self.mid_speed) / self.arrival_time  # Arrival phase slope

	def V1(self,t):
		return self.departure_speed + self.Cd * t  # Velocity on departure phase equation
	def V2(self,t):
		return self.mid_speed  # Velocity on max speed phase equation
	def V3(self,t):
		return self.arrival_speed + self.Ca * (t - self.total_time)  # Velocity on arrival phase equation

	def PlannedVelocity(self, t):
		if t<self.total_time:
			return min(self.V1(t), self.V2(t), self.V3(t))
		else:
			return self.arrival_speed

	def PlannedTotalTime(self):
		return self.total_time

	def PlannedPosition(self, t):
		# Phase 1: Departure
		if t <= self.departure_time:
			return self.departure_speed * t + 0.5 * self.Cd * t ** 2
		# Phase 2: Max speed
		elif t <= self.departure_time + self.mid_time:
			t_m = t - self.departure_time
			return self.departure_distance + self.mid_speed * t_m
		# Phase 3: Arrival
		elif t <= self.total_time:
			t_a = t - self.departure_time - self.mid_time
			return self.departure_distance + self.mid_distance + self.mid_speed * t_a + 0.5 * self.Ca * t_a ** 2
		# After planned time, position remains at final total_distance
		else:  
			return self.total_distance

File: robot1/teensy_moteur/Curve.cpp
#include <iostream>
#include <cmath>

class Curve
{
private:
	float Vi, Vm, Vf;

	float D, D1, D2, D3;

	float T, T1, T2, T3;

	float C1, C3;

public:
	Curve(float intial_velocity, float max_velocity, float final_velocity, float total_distance, float acceleration_distance, float deceleration_distance);
	~Curve() = default;
	float PlannedVelocity(float t);
	float PlannedTotalTime();
};

Curve::Curve(float intial_velocity, float max_velocity, float final_velocity, float total_distance, float acceleration_distance, float deceleration_distance)
{
	this->Vi = intial_velocity;
	this->Vm = max_velocity;
	this->Vf = final_velocity;

	this->D = total_distance;
	this->D1 = acceleration_distance;
	this->D3 = deceleration_distance;
	this->D2 = this->D - this->D1 - this->D3;

	this->T1 = 2 * this->D1 / (this->Vi + this->Vm); // total time of departure phase
	this->T3 = 2 * this->D3 / (this->Vf + this->Vm); // total time of arrival phase
	this->T2 = this->D2 / this->Vm;

	this->T = this->T1 + this->T2 + this->T3;

	this->C1 = (this->Vm - this->Vi) / this->T1; // Departure phase slope
	this->C3 = (this->Vf - this->Vm) / this->T3; // Arrival phase slope
}

float Curve::PlannedTotalTime()
{
	return this->T;
}

float Curve::PlannedVelocity(float t)
{
	if (t > this->PlannedTotalTime())
	{
		return this->Vf;
	}
	else
	{
		float V1 = this->Vi + this->C1 * t;
		float V2 = this->Vm;
		float V3 = this->Vf + this->C3 * (t - this->T);

		return std::min(std::min(V1, V2), V3);
	}
}

int main()
{
	Curve c = Curve(0, 10, 0, 1000, 100, 200);
	float t = 0.0f;
	while (t < c.PlannedTotalTime())
	{
		std::cout << (t == 0.0f ? "" : ",") << c.PlannedVelocity(t);
		t += 0.001f;
	}
	std::cout << std::endl
			  << c.PlannedTotalTime();
	return 0;
}
